recorder/src/main/cpp/perf_ctx.hh

```diff
@@ -0,0 +1,156 @@
+#include <cstdint>
+#include <stack>
+#include <stdexcept>
+#include <array>
+#include <vector>
+#include "util.hh"
+#include <mutex>
+#include <cuckoohash_map.hh>
+#include <city_hasher.hh>
+#include <concurrentqueue.h>
+#include "prob_pct.hh"
+#include <unordered_map>
+#include <functional>
+
+#ifndef PERF_CTX_H
+#define PERF_CTX_H
+
+namespace PerfCtx {
+    typedef std::uint64_t TracePt;
+    /*
+      Ctx bit layout: it is composed of 3 parts, coverage %, merge-semantic and ctx-id and merge-semantic is not relevant for merge-generated ctx
+      because they can never get pushed over some other context (other contexts may get pushed over them, but then the parent's merge-semantic
+      doesn't matter).
+          
+      | bits [0 - 63] = 64 | [63 - 63] = 1 | [62 - 56] = 7 | [55 - 53] = 3  | [52 - 13] = 40 | [12 - 0] = 13                   |
+      |--------------------+---------------+---------------+----------------+----------------+---------------------------------|
+      | user-created       |             0 | coverage %    | merge-semantic | **unused**     | ctx-id (always a prime number)  |
+      |--------------------+---------------+---------------+----------------+----------------+---------------------------------|
+
+      13 bits can house 100th prime number (7919, which is less than 8192), so definitely sufficient for forseeable future.
+
+      | bits [0 - 63] = 64 | [63 - 63] = 1 | [62 - 7] = 56               | [6 - 0] = 7                 |
+      |--------------------+---------------+-----------------------------+-----------------------------|
+      | merge-generated    |             1 | ctx-id part-0 (combination) | ctx-id part-1 (permutation) |
+      |--------------------+---------------+-----------------------------+-----------------------------|
+
+      This scheme allows us to pack all ctx-ids in jlong, hence allowing 0-lookup impl on perf-data processing path (sigprof handler, allocator-callback etc) while keeping things simple and fast. Obviously, it is possible to bump this limit by using 2 or 3 or more jlongs, but because this the final aggregated data cardinality needs to be such that its easy for humans to consume, such enhancements shoudldn't be necessary.
+
+      This allows us to offer upto 349 user defined ctx, because 2351 is the 349th prime number and value of 2357^5 (next prime, 350th) is greater than 2^56. Since we support 5-depth nesting (as 2351^5 < 2^56) we need 7 bits for storing permutation-id (5! = 120, the next power of 2 is 128, ie 2^7).
```

Comment:
This is actually right. Explained to Ankur how this works offline. Its basically this:

2351^5 < 2^56 < 2357^5

2351 is the 349th prime number, 2357 is the 350th one.